subprojects/core/src/main/java/org/gradle/internal/service/scopes/VirtualFileSystemServices.java

```diff
@@ -199,8 +200,12 @@ public void beforeComplete() {
             return locationsWrittenByCurrentBuild;
         }
 
-        WatchableFileSystemDetector createWatchableFileSystemDetector(FileSystems fileSystems) {
-            return new DefaultWatchableFileSystemDetector(fileSystems);
+        WatchableFileSystemDetector createWatchableFileSystemDetector(FileSystems fileSystems, NativeCapabilities nativeCapabilities) {
+            if (nativeCapabilities.useFileSystemWatching()) {
+                return new DefaultWatchableFileSystemDetector(fileSystems);
+            } else {
+                return Stream::empty;
```

Comment:
This thing doesn't need to be a service at all. We could create it in its only call site: in `FileSystemWatchingBuildActionRunner.run()`, and only if we `nativeCapabilities.useFileSystemWatching()` returns `true`.

Reply:
I introduced this since I didn't want to make a call to native platform when the platform doesn't even support file system watching. We deal with this in the virtual file system by creating a `WatchingNotSupportedVirtualFileSystem`.

I moved the whole unsupported file system detection now to `WatchingVirtualFileSystem`, so I don't need to make this check any more.